evm/src/keccak_sponge/keccak_sponge_stark.rs

```diff
@@ -0,0 +1,441 @@
+use std::borrow::Borrow;
+use std::iter;
+use std::marker::PhantomData;
+use std::mem::size_of;
+
+use itertools::Itertools;
+use plonky2::field::extension::{Extendable, FieldExtension};
+use plonky2::field::packed::PackedField;
+use plonky2::field::polynomial::PolynomialValues;
+use plonky2::field::types::Field;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::timed;
+use plonky2::util::timing::TimingTree;
+
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use crate::cpu::kernel::keccak_util::keccakf_u32s;
+use crate::cross_table_lookup::Column;
+use crate::keccak_sponge::columns::*;
+use crate::memory::segments::Segment;
+use crate::stark::Stark;
+use crate::util::{trace_rows_to_poly_values, u32_from_le_bits, u32_to_le_bits, u8_to_le_bits};
+use crate::vars::StarkEvaluationTargets;
+use crate::vars::StarkEvaluationVars;
+
+#[allow(unused)] // TODO: Should be used soon.
+pub(crate) fn ctl_looked_data<F: Field>() -> Vec<Column<F>> {
+    let cols = KECCAK_SPONGE_COL_MAP;
+    let outputs = Column::singles(&cols.updated_state_u32s[..KECCAK_RATE_U32S]);
+    Column::singles([
+        cols.context,
+        cols.segment,
+        cols.virt,
+        cols.timestamp,
+        cols.len,
+    ])
+    .chain(outputs)
+    .collect()
+}
+
+#[allow(unused)] // TODO: Should be used soon.
+pub(crate) fn ctl_looking_keccak<F: Field>() -> Vec<Column<F>> {
+    let input_rate_cols = (0..KECCAK_WIDTH_U32S)
+        .map(|i| Column::le_bits(&KECCAK_SPONGE_COL_MAP.original_rate_bits[i * 32..(i + 1) * 32]));
+    let input_capacity_cols = Column::singles(
+        (0..KECCAK_WIDTH_U32S).map(|i| KECCAK_SPONGE_COL_MAP.original_capacity_u32s[i]),
+    );
+    let output_cols = Column::singles(
+        (0..KECCAK_WIDTH_U32S).map(|i| KECCAK_SPONGE_COL_MAP.updated_state_u32s[i]),
+    );
+    input_rate_cols
+        .chain(input_capacity_cols)
+        .chain(output_cols)
+        .collect()
+}
+
+#[allow(unused)] // TODO: Should be used soon.
+pub(crate) fn ctl_looking_memory<F: Field>(i: usize) -> Vec<Column<F>> {
```

Comment:
Hm I'm not sure I understand how this is supposed to work. It seems like this is the CTL columns for the `i`th byte in the block, but it always looks up the same address at `virt` regardless of `i`. 
Should the address be `virt + already_absorbed_bytes + i` instead?

Reply:
Ah yes, good catch.